Regex support added for ignore URL list in settings.py

sauravsharma1998 opened the pull request:
Regex Support Added for IGNORE_URLS:
admin/* [All URL pattern with starting with admin/ will be ignored for GUID generation]

Suggestion Link: https://github.com/snok/django-guid/issues/104

sauravsharma1998 commented:
Can someone please review the installation failure for **Django 5.0**
<img width="1410" alt="image" src="https://github.com/snok/django-guid/assets/80733574/95ba08cd-d6ea-40fe-beab-bd6b29996b6e">

sauravsharma1998 commented:
@JonasKs @nschlemm Can you please review workflows for **Django Version 5.0** above.

JonasKs commented:
Thanks for the PR!  
Don't worry about the pipelines, I would look into them before merging. 

The code needs tests, and I'd like to use a `re.compile` solution instead of this. Inspiration can be found [here](https://github.com/snok/django-auth-adfs/blob/378f14129d774ac035804e09f7b6a1b3f5a3f71f/django_auth_adfs/middleware.py#L19), and the check can be found in the same file, on line 45. 😊

sauravsharma1998 commented:
@JonasKs Thanks for your recommendations.

Followed this - The code needs tests, and I'd like to use a re.compile solution instead of this. Inspiration can be found [here](https://github.com/snok/django-auth-adfs/blob/378f14129d774ac035804e09f7b6a1b3f5a3f71f/django_auth_adfs/middleware.py#L19)

Updates: 
1. Fixed one previous test case where monkey patching was missing for updating the django settings
2. Added the test case for regex support for IGNORE_URL

Please have a look on the Django 5.0 workflows.

sauravsharma1998 commented:
@JonasKs @nschlemm @ingvaldlorentzen Can you please review the PR.

JonasKs commented:
I'll get to it when I have time. I won't have time until next week at earliest.  
This isn't paid work, it's done in our spare time.

sauravsharma1998 commented:
@JonasKs No issues. Check this as per your convenience. Thanks 👍

JonasKs commented:
Wow, I definitely didn't intend to just close this out of nowhere. I was looking for this now, and couldn't find it. 

@sauravsharma1998 , I'm sorry.

JonasKs commented:
The implementation would be a breaking change - since you change the current `ignore_urls` behavior. This should be a completely new setting.

JonasKs commented on django_guid/utils.py:
```diff
@@ -91,3 +92,22 @@ def validate_guid(original_guid: str) -> bool:
+    IGNORE_URLS = []
+    for url in settings.ignore_urls:
+        url_regex = url.replace('*', r'[\s\S]*')  # noqa
+        url_regex = '^' + url_regex + '$'
```
If we make this a new setting, e.g. `ignore_url_regex`, we can also skip this.

sauravsharma1998 replied:
@JonasKs Thanks for the suggestions. I have moved it to a new settings. But facing errors in CI pipeline.

JonasKs replied:
I’m on vacation, but I’ll check when I’m back. 😊

JonasKs commented:
There's failing pipelines, even though I haven't changed anything and I've deleted all caches.. I'll have to look further into this, but I'm not sure what's happening.

hartungstenio commented on django_guid/utils.py:
```diff
@@ -91,3 +92,20 @@ def validate_guid(original_guid: str) -> bool:
+    for url in settings.ignore_regex_urls:
+        url_regex = url.replace('*', r'[\s\S]*')  # noqa
+        url_regex = '^' + url_regex + '$'
+        IGNORE_URLS.append(re.compile(url_regex))
```
This will compile every regex twice every requests. It seems like a waste of processing power, and we will not get the benefits of compiled regex since they will bê compiled every time they are needed.

I wonder if it is possible to compile tem just once, maybe in the middleware setup or at the `ignore_regex_urls` settings.

hartungstenio commented on django_guid/utils.py:
```diff
@@ -91,3 +92,20 @@ def validate_guid(original_guid: str) -> bool:
+        url_regex = url.replace('*', r'[\s\S]*')  # noqa
+        url_regex = '^' + url_regex + '$'
+        IGNORE_URLS.append(re.compile(url_regex))
+    return any(url.match(endpoint) for url in IGNORE_URLS)
```
```suggestion
    return any(url.search(endpoint) for url in IGNORE_URLS)
```

Using `search` would remove the needs to change the user's regex a few lines above, and would make it compatible with the way django itself handle this kind of settings, like `SECURE_REDIRECT_EXEMPT‎`

hartungstenio reviewed (commented):
Could you update the README with the New behaviour?

JonasKs commented:
I'll close this for being stale, but thanks @hartungstenio for the review and @sauravsharma1998 for the work of this. If anyone wants to create a new PR with the reviewed changes, please do. 😊
